skimage/segmentation/random_walker_segmentation.py

```diff
@@ -263,9 +264,12 @@ def _preprocess(labels):
     return labels, nlabels, mask, inds_isolated_seeds, isolated_values
 
 
-def random_walker(data, labels, beta=130, mode='cg_j', tol=1.e-3, copy=True,
-                  multichannel=False, return_full_prob=False, spacing=None,
-                  *, prob_tol=1e-3):
+def random_walker(data: ImageArray, labels: LabelsArray, beta: float = 130,
+                  mode: Literal['cg', 'cg_j', 'cg_mg', 'bf'] = 'cg_j',
+                  tol: float = 1.e-3, copy: bool = True,
+                  multichannel: bool = False, return_full_prob: bool = False,
+                  spacing: Optional[Sequence] = None, *,
+                  prob_tol: float = 1e-3) -> LabelsArray:
```

Comment:
```suggestion
def random_walker(
    data: ImageArray,
    labels: LabelsArray,
    beta: float = 130,
    mode: Literal['cg', 'cg_j', 'cg_mg', 'bf'] = 'cg_j',
    tol: float = 1.e-3,
    copy: bool = True,
    multichannel: bool = False,
    return_full_prob: bool = False,
    spacing: Optional[Sequence] = None,
    *,
    prob_tol: float = 1e-3,
) -> LabelsArray:
```

Reply:
Does `Literal` mean that I can't do:

```python
mode = get_mode_from_gui_button()
output = random_walker(image, labels, mode=mode)
```
?

Reply:
Would you type your `get_mode_from_gui_button` or not? If you type it the script above should pass `mypy`. But probably you see `get_mode_from_gui_button` as a more generic function, and then it would not be possible to type it with such a specific type as `Literal['cg', 'cg_j', 'cg_mg', 'bf']`?

Reply:
@jni could you please explain in more details what you meant here? Remember that type hints are not used at runtime but only by the type checker.

Reply:
@emmanuelle yes but that is perfectly valid code to run through the type checker! I meant with my code snippet to show an example in which `mode=` would be a variable, not always a literal string.

Also side note that function type annotations *are* available at runtime in the `__annotations__` attribute of the function object!

Reply:
So I thought that `Literal` was a cool thing in particular for UI generation (dropdown with the discrete values) but it is quite restrictive for type checking indeed. For example
```python
from typing_extensions import Literal
from typing import Any


def my_func(x: Literal['a', 'b']) -> str:
    return x + x


x: Any = 'a'
my_func(x) # ok
y: Any = 'c'
my_func(y) # ok even if wrong value, but with type Any it passes
z: str = 'a'
my_func(z) # not ok because type is str and not Literal
```

Reply:
> (dropdown with the discrete values) 

Perhaps another good use for 'Annotated'? =)

Reply:
hum yes very smart! I must admit I don't completely understand how `Annotated` works (even after reading PEP 593). For example in the following code it's not possible to write directly `Annotated[str, 'a', 'b']` in the function signature and I don't understand why. There are workarounds like defining the annotated type outside of the signature, so it's not a big deal, but I'd prefer to understand this better...
```python
from typing_extensions import Annotated

def modes():
    return ['a', 'b']

# error: Name 'a' is not defined
# error: Name 'b' is not defined
def my_func(x: Annotated[str, "a", "b"]) -> str:
    return x + x

# ok
def my_other_func(x: Annotated[str, modes()]) -> str:
    return x + x

# ok
modes_type = Annotated[str, 'a', 'b']
# ok
def yet_another_func(x: modes_type) -> str:
    return x + x
```

Reply:
I think you are mixing up some concepts here. `Literal` is a type, so it can also be a type of a variable. It just means only specific values are allowed. In this case it would mean that `get_mode_from_gui_button()` needs to have a return type of `Literal['cg', 'cg_j', 'cg_mg', 'bf']` (or possibly a subset of those values) to guarantee that only those values will be passed to `random_walker`

I think the most intuitive explanation I can come up with is that Literals are like Enums, but without the members having a name or the enum itself having a name.

In the case of `Literal['cg', 'cg_j', 'cg_mg', 'bf'] = 'cg_j'` an Enum does seem like the more appropriate choice though. Which has the advantage that the meaning of the values can be defined and documented in one place, and every place it is used in can then also be found (by simply just looking for code that imports this enum)

The Annotated feature is something I haven't used myself yet. I think the intention for it are third party extensions for complex type checking like dependent types ( which is feature that is essential to properly type numeric code like anything that uses numpy, but that is a different topic). My intuition is that it is most likely not the solution to whatever problem you are trying to solve, but I don't fully understand the problem itself yet.

Reply:
@fmagin 

> I don't fully understand the problem itself yet

I'm glad you asked! 😂 We have two issues:
- NumPy arrays do double and triple duty in the library: they can represent an image (which can be 2D, 3D, nD, 2D+channels, ...), a set of coordinates, a set of spatial extents... So we need a way to distinguish between all these types, while still allowing bare NumPy arrays through no problem. Here NewType is indeed what I thought was the right solution, although it is not the right solution e.g. for distinguishing between multichannel and single channel arrays.
- The other issue we have is that for many functions (and we hope to continue to increase how many), not just NumPy arrays are allowed, but actually any array-like, if they implement `__array__` and maybe NEP-18 (`__array_function__`). In this case, NewType falls over because we actually need to use a protocol, and that protocol is defined by NumPy, and you can't use NewType with a protocol.